### ISSUE TITLE
Add constraint/objective construction from NumPy arrays

### ISSUE THREAD

**Issue by arcondello** [dwavesystems/dimod#1308]
Requested by @mhramani https://github.com/dwavesystems/dimod/pull/1307#pullrequestreview-1276064366.

Something like
```python
indices = [0, 1, 5]
biases = [1, 2, 3]
cqm.add_constraint((indices, biases), "==", 5)
```
which would be interpreted as a linear constraint $x_0 + 2x_1 + 3x_5 = 5$.

**Comment by ACE07-Sev** [comment]
Like so?

```
from dimod import ConstrainedQuadraticModel, Binary, quicksum

cqm_test = ConstrainedQuadraticModel()
vars = [Binary(f"X{i}") for i in range(7)]

def add_constraint(cqm, vars, indices, biases, sense, rhs):
    if len(indices) != len(biases):
        raise ValueError('The indices and biases do not have same size.')
    constraint = quicksum(biases[i] * vars[indices[i]] for i in range(len(indices)))
    cqm.add_constraint(constraint, sense=sense, rhs=rhs, label='c')

add_constraint(cqm_test, vars, [0, 2, 5], [1, 2, 3], "==", 5)

print(cqm_test.constraints['c'].to_polystring())
>> X0 + 2*X2 + 3*X5 == 5.0
```

**Comment by arcondello** [comment]
Yes, something like that. Though the performance benefits of such a thing would come from implementing it at the Cython level, likely as a sibling method to [cyConstrainedQuadraticModel.add_constraint_from_iterable()](https://github.com/dwavesystems/dimod/blob/e5a8e3b127ab0c7f86eb9f7d57fc0b29a34b984e/dimod/constrained/cyconstrained.pyx#L145) and [cyConstrainedQuadraticModel.add_constraint_from_model()](https://github.com/dwavesystems/dimod/blob/e5a8e3b127ab0c7f86eb9f7d57fc0b29a34b984e/dimod/constrained/cyconstrained.pyx#L176).

**Comment by matti-sarjala** [comment]
A useful design boundary here may be to keep the array interface explicit about when raw numeric data becomes a CQM expression.

For users coming from NumPy / scipy / Pyomo-style workflows, an array representation is natural, but it can be ambiguous unless the contract fixes:

1. variable ordering,
2. index-to-variable mapping,
3. coefficient dtype / numeric coercion,
4. linear vs quadratic interpretation,
5. sense and rhs handling,
6. duplicate indices.

Possible acceptance criteria:

- `indices` and `biases` must have matching length,
- duplicate indices are either summed or rejected explicitly,
- out-of-range indices give a clear error,
- integer variables / binary variables use the provided variable list or ordering consistently,
- the resulting constraint has the same energy as the equivalent expression built with `quicksum`,
- objective construction and constraint construction use the same index/coefficient convention.

This would make the feature useful for high-throughput model construction while avoiding a silent mismatch between numeric array position and optimization-variable identity.
